content/posts/11-alpine-dev-env-in-fedora.md

```diff
@@ -0,0 +1,644 @@
+    alpine-dev-env-for-%u \
+    sh -c "echo '%u ALL=(ALL) NOPASSWD: ALL' >> /etc/sudoers.d/%u-all-nopasswd"
+# Add local repos
+ExecStartPost=%%PODMAN%% exec \
+    alpine-dev-env-for-%u \
+    sh -c "printf -- >> /etc/apk/repositories '%h/packages/%%s\n' main community testing"
+```
+
+Someone already commented so lets go to the next one.
+
+```ini
+ExecStop=%%PODMAN%% stop \
+    --time 2 \
+    --cidfile %t/alpine-dev-env-for-%u.ctr-id
+ExecStopPost=%%PODMAN%% rm \
+    --ignore \
+    --force \
+    --cidfile %t/alpine-dev-env-for-%u.ctr-id
+```
+
+When stopping, wait 2 seconds before we force our way in, after stopping the service, delete the container forcefully.
+
+```ini
+PIDFile=%t/alpine-dev-env-for-%u.pid
+KillMode=none
+Type=forking
+
+[Install]
+WantedBy=multi-user.target
+```
+
+- `PIDFile` uses the one created by `podman run` in `ExecStart`
+- `KillMode` is set to `none` because we bring down a container, not a service
+- `Type` is `forking` because we make podman `--detach` itself into the background
+
+## abuild
+
+We now need a wrapper for running abuild commands, it needs to perform certain actions before we run in the container.
+
+```sh
+#!/bin/sh
+die() {
+    printf "%s\\n" "$@" >&2
+    exit 1
+}
+
+run() {
+    ssh -p %%PORT%% -tt \
+        -o User=$USER \
+        -o IdentityFile=~/.ssh/alpine-dev-env \
+        -o NoHostAuthenticationForLocalhost=yes \
+        -o LogLevel=QUIET \
+        localhost "$@"
+}
+
+: ${APORTSDIR:=%%APORTSDIR%%}
+
+APORTSDIRNAME="$(basename $APORTSDIR)"
+
+## check running from within an `aports` tree
+if [ "${PWD%*/$APORTSDIRNAME/*}" = "$PWD" ]; then
+    die "Error: expecting to be run from within an aports tree!" \
+        "It must be run under '$APORTSDIR'"
+fi
+
+# Generate new abuild key if not set
+cmd="grep -sq '^PACKAGER_PRIVKEY=' $HOME/.abuild/abuild.conf || abuild-keygen -n -a"
+
+# Copy keys over, we need this so packages will actually succeed when
+# building (because of the signing stage), this is also required to
+# install them
+cmd="$cmd && sudo cp $HOME/.abuild/*.rsa.pub /etc/apk/keys"
+
+case "$1" in
+    checksum|unpack) cmd="$cmd; cd $APORTSDIR/${PWD#*/$APORTSDIRNAME/} && abuild $@" ;;
+    *) cmd="$cmd && cd $APORTSDIR/${PWD#*/$APORTSDIRNAME/} && sudo apk upgrade -U -a && abuild $@" ;;
+esac
+
+if ! systemctl --user is-active --quiet alpine-ssh.service; then
+    die "alpine-ssh.service must be active, run systemctl --user start alpine-ssh.service"
+fi
+
+run "$cmd"
+```
+
+Notice a few things:
+
+- We run `apk upgrade -U -a` for any command other than `checksum` and `unpack`, the latter don't build anything and won't get maluses from having an outdated system.
+- We use systemd to tell us if we are running the service
+
+## Makefile
+
+to bring it all together and provide a nice workflow for building the image and installing the files lets write a makefile.
+
+> Please don't use makefiles if you can help, use [meson][6] or redo (for simpliciy)
```

Comment:
I don't think this remark is necessary

Reply:
Yes, I was just scared I might be passing the idea that using Makefiles is in any shape, way or form ok